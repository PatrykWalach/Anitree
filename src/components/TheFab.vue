<template>
  <v-fab-transition>
    <v-btn
      v-if="!bottomNavigation"
      :key="active.icon"
      :color="active.color"
      :to="active.to"
      v-on="active.on"
      fab
      large
      dark
      fixed
      bottom
      right
    >
      <v-icon>{{ active.icon }}</v-icon>
    </v-btn>
  </v-fab-transition>
</template>
<script lang="ts">
import { SetupContext, computed, createComponent } from '@vue/composition-api'
import TheSearchFilters from './TheSearchFilters.vue'
import { useBottomNavigation } from '@/App.vue'

export const useFab = (root: SetupContext['root']) => {
  const active = computed(() => {
    const { filter, timeline, edit } = root.$modules

    switch (root.$route.name) {
      case 'search':
        return {
          color: 'accent',
          icon: 'tune',
          on: {
            click: () => filter.mutations.CHANGE_IS_SHOWN(true),
          },
        }
      case 'media-timeline':
        return {
          color: 'accent',
          icon: 'sort',
          on: {
            click: () =>
              timeline.mutations.CHANGE_ORDER(timeline.state.order.value * -1),
          },
        }
      case 'media-about':
        return {
          color: 'error',
          icon: 'edit',
          on: {
            click: () =>
              edit.actions.open(parseInt(root.$route.params.mediaId, 10)),
          },
        }
      default:
        return {
          color: 'primary',
          icon: 'search',
          to: { name: 'search', query: {} },
        }
    }
  })

  return {
    active,
    ...useBottomNavigation(root),
  }
}

export default createComponent({
  components: {
    TheSearchFilters,
  },
  setup(_, { root }) {
    return useFab(root)
  },
})
</script>
