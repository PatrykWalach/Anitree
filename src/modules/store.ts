import * as modules from '.'
import { Store, plugin } from 'vuex-composition-api'

export const store =
  process.env.NODE_ENV === 'production'
    ? undefined
    : new Store({
        plugins: [plugin(Object.values(modules))],
      })
