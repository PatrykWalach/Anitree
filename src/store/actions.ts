export { titleActions as title } from './modules/title'

export { editActions as edit } from './modules/edit'
export { filterActions as filter } from './modules/filter'
export { shareActions as share } from './modules/share'
export { settingsActions as settings } from './modules/settings'
export { timelineActions as timeline } from './modules/timeline'
export { changesActions as changes } from './modules/changes'
