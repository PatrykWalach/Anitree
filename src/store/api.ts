import gql from 'graphql-tag'
import { FetchVariables } from './models'

import apollo from '../apollo'
import { AMedia } from '@/types'

export const query = gql`
  query($search: String, $idIn: [Int], $page: Int = 0) {
    Page(page: $page, perPage: 50) {
      media(search: $search, id_in: $idIn) {
        id
        bannerImage
        coverImage {
          extraLarge
          large
          medium
          color
        }
        format
        status
        tags {
          name
          id
          rank
        }
        siteUrl
        type
        studios(isMain: true) {
          nodes {
            name
            id
          }
        }
        startDate {
          year
          month
          day
        }
        endDate {
          year
          month
          day
        }
        title {
          romaji
          english
          native
        }
        description
        mediaListEntry {
          status
        }
        relations {
          edges {
            node {
              id
              type
            }
            id
            relationType
          }
        }
      }
    }
  }
`

export function fetchMediaApollo(variables: FetchVariables): Promise<AMedia[]> {
  return apollo
    .query<{ Page: { media: AMedia[] } }>({
      query,
      variables
    })
    .then(({ data }) => data.Page.media)
}
