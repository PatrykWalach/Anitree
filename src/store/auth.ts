import VueFunctionApi, { ref, Ref, computed } from 'vue-function-api'
import Vue from 'vue'
Vue.use(VueFunctionApi)

import { Token } from '@/types'
const stored = localStorage.getItem('AUTH')

const auth: Ref<Token | null> = ref((stored && JSON.parse(stored)) || null)

const token = computed(() => {
  if (
    auth.value
    // && Date.now() < parseInt(auth.value.expires_in)
  ) {
    return auth.value.access_token
  }
  return null
})

const CHANGE_TOKEN = async (payload: Token | null) => {
  if (payload) {
    const newAuth: Token = {
      ...payload,
      /* eslint-disable-next-line @typescript-eslint/camelcase */
      expires_in: (Date.now() + parseInt(payload.expires_in)).toString()
    }
    localStorage.setItem('AUTH', JSON.stringify(newAuth))
    auth.value = newAuth
  } else {
    auth.value = payload
  }
}

export default function useAuth() {
  return { auth, CHANGE_TOKEN, token, stored }
}
