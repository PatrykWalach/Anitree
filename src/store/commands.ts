import CompositionApi, { Ref, ref, watch } from '@vue/composition-api'
import {
  DeleteCommand,
  DeleteCommandConstructor,
} from './commands/DeleteCommand'
import { DeleteVariables, SaveVariables } from '@/graphql/schema/listEntry'
import { SaveCommand, SaveCommandConstructor } from './commands/SaveCommand'
import { Form } from '@/types'
import { FuzzyDate } from '@/graphql/schema/media'
import { MediaList } from '@/graphql/schema/mediaListCollection'
import Vue from 'vue'
import { useSettings } from './settings'

Vue.use(CompositionApi)

export interface Command {
  saveState?(): void | Promise<void>
  undo?(): void | Promise<void>
  execute(): void | Promise<void>
}

export interface ListCommand extends Command {
  backup: Form | null
  variables: SaveVariables | (DeleteVariables & Pick<SaveVariables, 'mediaId'>)
  done: boolean
  _class: 'SaveCommand' | 'DeleteCommand'
}

export const isDeleteCommand = (
  e: (SaveCommandConstructor | DeleteCommandConstructor) &
    Pick<ListCommand, '_class'>,
): e is DeleteCommandConstructor & Pick<DeleteCommand, '_class'> =>
  e._class === 'DeleteCommand'

export const isSaveCommand = (
  e: (SaveCommandConstructor | DeleteCommandConstructor) &
    Pick<ListCommand, '_class'>,
): e is SaveCommandConstructor & Pick<SaveCommand, '_class'> =>
  e._class === 'SaveCommand'

const stored = localStorage.getItem('CHANGES')

const history: Ref<ListCommand[]> = ref(
  (stored &&
    JSON.parse(stored, (key, value) => {
      if (value instanceof Object) {
        if (isSaveCommand(value)) {
          return new SaveCommand(value)
        }
        if (isDeleteCommand(value)) {
          return new DeleteCommand(value)
        }
      }
      return value
    })) ||
    [],
)

watch(() => {
  if (useSettings().cacheChanges.value) {
    localStorage.setItem('CHANGES', JSON.stringify(history.value))
  }
})
;(async () => {
  for (const command of history.value.filter(({ done }) => !done)) {
    await command.execute()
  }
})()

const dispatch = async (command: ListCommand) => {
  history.value.push(command)
  await command.execute()
  return command
}

export const useCommands = () => {
  return {
    dispatch,
    history,
  }
}

export const mediaListToForm = (
  mediaListEntry: MediaList | null,
  advancedScoring: string[],
): Form => {
  if (mediaListEntry) {
    const advancedScores = advancedScoring
      .map(key => mediaListEntry.advancedScores[key])
      .map(value => value || 0)

    delete mediaListEntry.__typename
    delete mediaListEntry.startedAt.__typename
    delete mediaListEntry.completedAt.__typename
    delete mediaListEntry.private
    delete mediaListEntry.id
    delete mediaListEntry.mediaId

    return {
      ...mediaListEntry,
      advancedScores,
    }
  }

  const advancedScores = advancedScoring.map(() => 0)
  return {
    advancedScores,
    completedAt: { day: null, month: null, year: null },
    notes: '',
    progress: 0,
    progressVolumes: 0,
    repeat: 0,
    score: 0,
    startedAt: { day: null, month: null, year: null },
    status: null,
  }
}

export const isDate = (e: any): e is FuzzyDate => e.__typename === 'FuzzyDate'
