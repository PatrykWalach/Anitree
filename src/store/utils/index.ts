export * from './useCommands'
export * from './useTitle'
export * from './useEdit'
