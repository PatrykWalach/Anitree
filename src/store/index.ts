import Vue from 'vue'
import Vuex from 'vuex'

// import { getModule } from 'vuex-module-decorators'
Vue.use(Vuex)

import { ModuleMedia } from './modules/media'
import { ModuleTheme } from './modules/theme'
import { ModuleTitle } from './modules/title'
import { ModuleAuth } from './modules/auth'

export interface RootState {
  media: ModuleMedia
  theme: ModuleTheme
  title: ModuleTitle
  auth: ModuleAuth
}

export default new Vuex.Store<RootState>({})
