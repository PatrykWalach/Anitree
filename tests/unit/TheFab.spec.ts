import { createVariations, matchSnapshot } from './utils'
import TheFab from '@/components/TheFab.vue'
import { useMockedStore } from './mocks/store'

describe('TheFab', () => {
  createVariations({
    mocks: {
      $route: () => [
        ...createVariations({
          name: () => ['search', 'home'],
        }),
        ...createVariations({
          name: () => ['media-timeline', 'media-about'],
          params: {
            mediaId: 1,
          },
        }),
      ],
    },
  }).forEach(settings =>
    matchSnapshot(TheFab, {
      ...settings,
      ...useMockedStore(),
    }),
  )
})
