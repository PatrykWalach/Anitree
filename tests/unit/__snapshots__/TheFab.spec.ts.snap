// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TheFab matches snapshot 1`] = `
<v-fab-transition-stub
  mode="out-in"
  origin="center center"
>
  <v-btn-stub
    activeclass=""
    color="accent"
    fab="true"
    large="true"
    right="true"
    tag="button"
    type="button"
  >
    <v-icon-stub
      color="white"
    >
      tune
    </v-icon-stub>
  </v-btn-stub>
</v-fab-transition-stub>
`;

exports[`TheFab matches snapshot 2`] = `
<v-fab-transition-stub
  mode="out-in"
  origin="center center"
>
  <v-btn-stub
    activeclass=""
    color="accent"
    exact="true"
    fab="true"
    large="true"
    right="true"
    tag="button"
    to="[object Object]"
    type="button"
  >
    <v-icon-stub
      color="white"
    >
      search
    </v-icon-stub>
  </v-btn-stub>
</v-fab-transition-stub>
`;

exports[`TheFab matches snapshot 3`] = `
<v-fab-transition-stub
  mode="out-in"
  origin="center center"
>
  <v-btn-stub
    activeclass=""
    color="accent"
    fab="true"
    large="true"
    right="true"
    tag="button"
    type="button"
  >
    <v-icon-stub
      color="white"
    >
      sort
    </v-icon-stub>
  </v-btn-stub>
</v-fab-transition-stub>
`;

exports[`TheFab matches snapshot 4`] = `
<v-fab-transition-stub
  mode="out-in"
  origin="center center"
>
  <v-btn-stub
    activeclass=""
    color="accent"
    fab="true"
    large="true"
    right="true"
    tag="button"
    type="button"
  >
    <v-icon-stub
      color="white"
    >
      edit
    </v-icon-stub>
  </v-btn-stub>
</v-fab-transition-stub>
`;
